bitcoin/src/network/constants.rs

```diff
@@ -92,22 +90,115 @@ impl Network {
     /// # Examples
     ///
     /// ```rust
-    /// use bitcoin::network::constants::Network;
+    /// use bitcoin::network::constants::{Network, Magic};
     ///
     /// let network = Network::Bitcoin;
-    /// assert_eq!(network.magic(), 0xD9B4BEF9);
+    /// assert_eq!(network.magic(), Magic::from_bytes([0xF9, 0xBE, 0xB4, 0xD9]));
     /// ```
-    pub fn magic(self) -> u32 {
-        // Note: any new entries here must be added to `from_magic` above
+    pub fn magic(self) -> Magic {
+        Magic::from(self)
+    }
+}
+
+/// Network magic bytes to identify the cryptocurrency network the message was intended for.
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub struct Magic([u8; 4]);
+
+impl Magic {
+    /// Create network magic from bytes.
+    pub fn from_bytes(bytes: [u8; 4]) -> Magic {
+        Magic(bytes)
+    }
+
+    /// Get network magic bytes.
+    pub fn to_bytes(self) -> [u8; 4] {
+        self.0
+    }
+}
+
+/// Error in parsing magic from string.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MagicError {
+    /// Magic string is invalid
+    InvalidStr(String),
+    /// The magic was not from a known network
+    UnknownMagic(Magic)
+}
+
+impl FromStr for Magic {
+    type Err = MagicError;
+
+    fn from_str(s: &str) -> Result<Magic, Self::Err> {
+        match u32::from_str_radix(s.trim_start_matches("0x"), 16) {
+            Ok(bytes) => Ok(Magic(bytes.to_be_bytes())),
+            Err(_) => Err(MagicError::InvalidStr(s.to_owned()))
+        }
+    }
+}
+
+impl From<Network> for Magic {
+    fn from(network: Network) -> Magic {
+        match network {
+            // Note: any new network entries must be added below
```

Comment:
This comment looks redundant but it'd make sense if it said "don't forget to add an entry to `try_from` below". Or better yet, use a macro to implement both so they can't be out of sync.

Reply:
Left as a comment for now but having a macro sounds like a good idea.